core-bundle/src/Resources/contao/library/Contao/Config.php

```diff
@@ -244,37 +247,48 @@ public function save()
 			$strFile .= "\n" . $this->strBottom . "\n";
 		}
 
-		$strTemp = md5(uniqid(mt_rand(), true));
+		$strTemp = Path::join(TL_ROOT, 'system/tmp', md5(uniqid(mt_rand(), true)));
 
 		// Write to a temp file first
-		$objFile = fopen(TL_ROOT . '/system/tmp/' . $strTemp, 'w');
+		$objFile = fopen($strTemp, 'w');
 		fwrite($objFile, $strFile);
 		fclose($objFile);
 
 		// Make sure the file has been written (see #4483)
-		if (!filesize(TL_ROOT . '/system/tmp/' . $strTemp))
+		if (!filesize($strTemp))
 		{
 			\System::log('The local configuration file could not be written. Have your reached your quota limit?', __METHOD__, TL_ERROR);
 
 			return;
 		}
 
+		$fs = new Filesystem();
+
 		// Adjust the file permissions (see #8178)
-		$this->Files->chmod('system/tmp/' . $strTemp, \Config::get('defaultFileChmod'));
+		$fs->chmod($strTemp, self::get('defaultFileChmod'));
+
+		// Create the destination path of the localconfig.php
+		$strDestination = Path::join(TL_ROOT, 'system/config/localconfig.php');
+
+		// Get the realpath in case it is a symlink
+		if ($fs->exists($strDestination))
+		{
+			$strDestination = realpath($strDestination);
+		}
 
 		// Then move the file to its final destination
-		$this->Files->rename('system/tmp/' . $strTemp, 'system/config/localconfig.php');
+		$fs->rename($strTemp, $strDestination, true);
 
 		// Reset the Zend OPcache
 		if (\function_exists('opcache_invalidate'))
 		{
-			opcache_invalidate(TL_ROOT . '/system/config/localconfig.php', true);
+			opcache_invalidate($strDestination, true);
```

Comment:
Are you sure that this is correct? OPcache internally uses the realpath instead of the symlink?

Reply:
OPcache uses the realpath of a file internally as far as I know. So in the end it should not matter which path you invalidate.

Reply:
FYI: Yes it does use the realpath internally. Here's the piece of code responsible for that in case you care: https://github.com/php/php-src/blob/5fdabeb9695d008afcc967bd6e0c60e889fa6883/ext/opcache/ZendAccelerator.c#L1293